Refactoring aggregation

LucaMarconato opened the pull request:
Removing `id_key` from aggregation to simplify the APIs.

kevinyamauchi commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -221,53 +254,186 @@ def _aggregate_shapes(
+        "Incompatible choice: aggregating a categorical column with " "agg_func='mean'"
     )
+
+    ONES_COLUMN = "__ones_column"
+    AREAS_COLUMN = "__areas_column"
```
can you please add a comment explaining what these are for

kevinyamauchi reviewed (approved):
Hey @LucaMarconato ! Thanks for walking me through your PR today. As I mentioned, I don't have time to review in depth.  I given the tests and that you will be testing these functions in the Xenium-Visium vignette, I think we should just merge after you've take a pass to clean up a bit (remove commented code blocks, `##`, etc.)

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+def _parse_element(element: str | SpatialElement, sdata: SpatialData | None, str_for_exception: str) -> SpatialElement:
+    if not ((sdata is not None and isinstance(element, str)) ^ (not isinstance(element, str))):
+        raise ValueError(
+            f"To specify the {str_for_exception!r} spatial element element, please do one of the following: "
```
```suggestion
            f"To specify the {str_for_exception!r} spatial element, please do one of the following: "
```

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+    if not ((sdata is not None and isinstance(element, str)) ^ (not isinstance(element, str))):
+        raise ValueError(
+            f"To specify the {str_for_exception!r} spatial element element, please do one of the following: "
+            f"- either pass a SpatialElement to the {str_for_exception!r} parameter (and keep "
```
spatialElement her is capitalized and up not

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
-    *,
-    value_key: str | None = None,
-    agg_func: str | list[str] = "mean",
+    values_sdata: SpatialData | None = None,
```
I don't understand how a function can have all default arguments, at least one should have a mandatory argument no?

LucaMarconato replied:
true, `values` and `by` are always mandatory, I will change the order. I made them optional because I had put them after the optional values `values_sdata` and `by_sdata`

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+        The values to aggregate: if values_sdata is None, must be a SpatialElement, otherwise must be a string
+        specifying the name of the SpatialElement in values_sdata
+    by_sdata
+        Regions to aggregate by: if None, by must be a SpatialElement; if not None, by must be a string.
```
please use `by` when referring to the argument otherwise it's the same as literal by

LucaMarconato replied:
Done, also for `values`

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
     target_coordinate_system
         Coordinate system to transform to before aggregating.
+    fractions
+        Adjusts for partial areas overlap between regions in values and by.
```
also here, `values` and `by`

giovp replied:
and everywhere else really

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+        Additional details:
+
+             - default is fractions = False.
+             - when aggregating points this values must be left to False, as the points don't have area, thus
+             otherwise a table of zeros will be obtained;
```
remove thus

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+             - default is fractions = False.
+             - when aggregating points this values must be left to False, as the points don't have area, thus
+             otherwise a table of zeros will be obtained;
+             - for categorical values count and sum are equivalent when fractions = False, but when fractions = True
```
```suggestion
             - for categorical values, `count` and `sum` are equivalent when fractions = False, but when fractions = True
```

giovp replied:
and also fix in rest

LucaMarconato replied:
done

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
+             - when aggregating points this values must be left to False, as the points don't have area, thus
+             otherwise a table of zeros will be obtained;
+             - for categorical values count and sum are equivalent when fractions = False, but when fractions = True
+             count and sum are different: count would give unmeaningful results and so it's not allowed, while sum
```
```suggestion
             count and sum are different: count would give not meaningful results and so it's not allowed, while sum
```

giovp replied:
also in the other please fix

LucaMarconato replied:
done

giovp commented on src/spatialdata/_core/operations/aggregate.py:
```diff
@@ -1,148 +1,256 @@
     Returns
     -------
-    AnnData of shape (by.shape[0], values[id_key].nunique())])
+    Returns a SpatialData object with the "by" shapes as SpatialElement and a table with the aggregated values
```
please use `by`

giovp commented on src/spatialdata/_core/query/relational_query.py:
```diff
@@ -48,8 +99,186 @@ def _filter_table_by_elements(table: AnnData, elements_dict: dict[str, dict[str,
+    Returns
+    -------
+    DataFrame or ndarray with the values requested.
+
```
remove empty line,

giovp commented on src/spatialdata/_core/query/relational_query.py:
```diff
@@ -48,8 +99,186 @@ def _filter_table_by_elements(table: AnnData, elements_dict: dict[str, dict[str,
+                elif value_key in table.var_names:
+                    origins.append(_ValueOrigin(origin="var", is_categorical=False, value_key=value_key))
+
+    # adding from the channel names of images
+    # TODO
```
remove

LucaMarconato replied:
done

giovp commented on src/spatialdata/_core/spatialdata.py:
```diff
@@ -213,91 +213,51 @@ def query(self) -> QueryManager:
-            Additional keyword arguments to pass to :func:`xrspatial.zonal_stats`.
+        Notes
+        -----
+        This function calls :func:`spatialdata.aggregate` with the convenience that values and by can be string
+        without having to specify the values_sdata and by_sdata, which in that case will be replaced by `self`.
```
by reading this docstring, why is values_sdata and by_sdata still there?

LucaMarconato replied:
@giovp because one may still want to aggregate values from a different object.

giovp replied:
yeah but this is the method right? by this logic there is no differentiation between what the function does and what the method does?

LucaMarconato replied:
Exactly, there is no differentiation and the method is a convenience. I am in favor on this as we do it also for `bounding_box_query`, `polygon_query` and `transform` (well in this case it's called `transform_to_coordinate_system`, but after the refactoring will also be renamed). I will push soon the notebook, there I use mostly the method because since all the elements are in the `sdata` objects there is no need to pass `values_sdata` and `by_sdata`, and I can just pass `values` and `by` as strings.

giovp commented on tests/conftest.py:
```diff
@@ -392,29 +409,47 @@ def _make_sdata_for_testing_querying_and_aggretation() -> SpatialData:
+        # otherwise pre-commits will remove the "unused" import
+        _ = spatialdata_plot
+
+        ax = plt.gca()
+        sdata.pl.render_shapes(element="values_polygons", color="categorical_in_obs").pl.show(ax=ax)
+        sdata.pl.render_shapes(element="values_circles", color="categorical_in_obs").pl.show(ax=ax)
+        sdata.pl.render_shapes(element="by_polygons", na_color=(1.0, 0.7, 0.7, 0.3)).pl.show(ax=ax)
+        sdata.pl.render_shapes(element="by_circles", na_color=(1.0, 0.7, 0.7, 0.3)).pl.show(ax=ax)
+        sdata.pl.render_points(color="categorical_in_ddf", size=10.0, palette="tab10").pl.show(ax=ax)
+        plt.show()
     return sdata
```
I don't think we want to have data plotting to be a dependency of this, I would remove this

LucaMarconato replied:
agree, we don't to have it as a dependency. It was useful for developing the tests, but I will move it to the sandbox.

giovp reviewed (commented):
mostly comment on docstrings. I must sa y I still don't like the values and sdata_values logic, but I guess it's too late to discuss again API

giovp reviewed (approved):
mostly comment on docstrings. I must sa y I still don't like the values and sdata_values logic, but I guess it's too late to discuss again API

giovp commented:
@LucaMarconato #303 introduced the pyupgrade typing so that stuff like `Optional[str]` should be changed to `str | None`.

I ignore such change selectively in some files (e.g. transformation module) and the `UP` code (pyupgrade rules by ruff) is added as `unfixable` (ruff can fix everything but I would keep this as unfixable as it can be useful to reason about it, especially be contributors). 

It will probably throw a bunch of errors in the files you are working on, so an option is to remove `UP` from unficable, run ruff with `pre-commit run --all-files` and then re-add `UP` to unfixable. So it fixes the files you want to get fixed automatically, but it won't in the future. 

I apologize for the conflicts but I took a look and they are very minor.

on a separate note, there are a lot of `per-file-ignore` in the transformation module. Would keep an eye on that when the transformation is refactored

LucaMarconato commented:
Thanks for the update! I will keep this in mind and gradually fix that!

LucaMarconato commented:
> mostly comment on docstrings. I must sa y I still don't like the values and sdata_values logic, but I guess it's too late to discuss again API

Thanks for the review, I now implemented the changes (only one open comment left). I actually think that the `value_sdata` and `by_sdata` usage is quite flexible and intuitive. I have updated the aggregation notebook so we can collect some feedback. Anyway, up to change/improve it later on!
